Trees/BinarySearchTree.cpp

```diff
@@ -0,0 +1,358 @@
+    two node, none of which are itself or its predecessors, i.e the graph is
+    acyclic. The first node is called the root node. Another important feature
+    of this tree is that each node has values lesser than or equal to itself in
+    the left subtree and values greater than or equal to itself in the right
+    subtree.
+*/
+
+#include "BinarySearchTree.hpp"
+#include <iostream>
+#include <climits>
+#include <stack>
+#define INF INT_MAX
+
+using namespace std;
+
+Node* get_node() {
+    Node* newNode = new Node();
+    newNode->left_child = newNode->right_child = nullptr;
+
+    return newNode;
+}
+
+BinarySearchTree::BinarySearchTree() {
+    root = get_node();
+    root->value = INF;
+}
+
+/*
+    Insert
+    This traverses the tree to find the right spot to insert the newly given
+    value. At each node, it checks if the node has a value lesser than or equal
+    to itself, if so it goes to its left subtree. Otherwise it goes to the right
+    subtree. When it meets an end point i.e. NULL it inserts the new Node along
+    with the new value
+
+    Time complexity:
+    Best case : O(log n), where n is the number of nodes in the tree
+    Worst case : O(n)
+
+    Space complexity:
+    O(1)
+*/
+
+bool BinarySearchTree::insert_node(int value) {
+    // If first insertion, replace the value of the already created node
+    if(root->value == INF) {
+        root->value = value;
+        return true;
+    }
+
+    Node* current = root;
+    Node* newNode = get_node();
+
+    // Failed to allocate memory to a new node
+    if (newNode == nullptr) {
+        return false;
+    }
+
+    while (true) {
+        if (value <= current->value) {
+            if (current->left_child == nullptr) {
+                current->left_child = newNode;
+                current->left_child->value = value;
+                return true;
+            } else {
+                current = current->left_child;
+            }
+        } else {
+            if (current->right_child == nullptr) {
+                current->right_child = newNode;
+                current->right_child->value = value;
+                return true;
+            } else {
+                current = current->right_child;
+            }
+        }
+    }
+}
+
+/*
+    Remove
+    This traverses the tree to find the given value. At each node, it checks if
+    the node has a value equal to the given value, if so it finds the inorder
+    successor of the node, and replaces the value to be deleted with that number.
+    Otherwise, it searches in the left subtree if the value is smaller than
+    node value and it searches the right subtree if the value is larger.
+
+    Time complexity:
+    Best case : O(log n), where n is the number of nodes in the tree
+    Worst case : O(n)
+*/
+
+bool BinarySearchTree::remove_node(int value) {
+    // If tree is empty remove is unsucessful
+    if (root->value == INF) {
+        return false;
```

Comment:
Simplify this function by breaking the different operations that happen in it (like finding the successor etc.) into separate functions, and call those functions here.